mlir/lib/Quantum/Transforms/remove_chained_self_inverse.cpp

```diff
@@ -0,0 +1,74 @@
+// Copyright 2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#define DEBUG_TYPE "chained-self-inverse"
+
+#include <memory>
+#include <vector>
+
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Errc.h"
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Bufferization/IR/Bufferization.h"
+#include "mlir/Dialect/Complex/IR/Complex.h"
+#include "mlir/Dialect/Index/IR/IndexDialect.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/SymbolTable.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "Catalyst/IR/CatalystDialect.h"
+#include "Quantum/IR/QuantumOps.h"
+#include "Quantum/Transforms/Patterns.h"
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst::quantum;
+
+namespace catalyst {
+namespace quantum {
+
+#define GEN_PASS_DEF_QUANTUMCANONICALIZE
+#include "Quantum/Transforms/Passes.h.inc"
+
+struct RemoveChainedSelfInversePass
```

Comment:
If we want to use this pass, do you see a good place in the code base where we could invoke it from?

Reply:
I'm sorry I don't know if I fully understand the question. I didn't think passes were invoked like a function call but intead registered and then run either in a pass pipeline or through command line arguments. 

If we wanted to use this pass, since it has been registered as a Catalyst Pass we should be able to run it using a command like `quantum-opt --remove-chained-self-inverse %input-file > %output-file`.

Reply:
You're right, the pass could be invoked via `quantum-opt`. I was thinking more in the context of the entire Catalyst package, as used by users to compile their Python programs with. No worries if not, but maybe you have some instincts about how we could make sure the optimization is applied when users compile their programs from Python?

Reply:
I see, hmm after looking through the codebase more I see that we have files like `CompilerDriver.cpp` and `PyCompilerDriver.cpp` and they seem like a good place to start to make sure the pass is applied when users use the compiler.

Reply:
Sounds good, thanks!